lxd/db/oidc/session.go

```diff
@@ -0,0 +1,240 @@
+package oidc
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/netip"
+	"time"
+
+	"github.com/google/uuid"
+	zitadelOIDC "github.com/zitadel/oidc/v3/pkg/oidc"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/lxd/lxd/auth/oidc"
+	"github.com/canonical/lxd/lxd/db"
+	"github.com/canonical/lxd/lxd/db/cluster"
+	"github.com/canonical/lxd/lxd/events"
+	"github.com/canonical/lxd/lxd/lifecycle"
+	"github.com/canonical/lxd/lxd/request"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+// NewSessionHandler returns a new [oidc.SessionHandler]. The getSessionExpiry function must return the current value
+// of oidc.session.expiry for the server configuration.
+func NewSessionHandler(db *db.Cluster, events *events.Server, getSessionExpiry func() string) oidc.SessionHandler {
+	return &sessionHandler{db: db, events: events, expiryFunc: getSessionExpiry}
+}
+
+type sessionHandler struct {
+	db         *db.Cluster
+	events     *events.Server
+	expiryFunc func() string
+}
+
+// StartSession starts a new session for the identity with the email given in the [oidc.AuthenticationResult].
+// For first time logins, a new identity is created. If the identity already exists, the identity metadata will be updated
+// to match the new values from the IdP. A session entry will be created in the database using details extracted from the
+// given [http.Request] and tokens.
+//
+// The [zitadelOIDC.Tokens] may be nil, this is when the client is authenticating with the CLI and their access tokens are stored locally.
+//
+// A [time.Time] can be provided to override the expiry of the session. This is used when a session is being refreshed to use
+// a newer cluster secret.
+func (s *sessionHandler) StartSession(r *http.Request, res oidc.AuthenticationResult, tokens *zitadelOIDC.Tokens[*zitadelOIDC.IDTokenClaims], expiryOverride *time.Time) (*uuid.UUID, *time.Time, error) {
+	// Get a new session UUID. This is a v7 UUID from which we can extract the session creation date.
+	sessionID, err := uuid.NewV7()
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create new session UUID: %w", err)
+	}
+
+	// Get the session expiry. Use the override, else get from config.
+	var expiry time.Time
```

Comment:
why not make the `expiry` var the argument to the function instead of `expiryOverride` and then if its nil then set it to the config one. Why the need to have separate var for expiryOverride?

Reply:
I wanted to make it clear that in general, the session handler is responsible for setting the expiry, but the caller can override this only when necessary.